Userland/Libraries/LibWeb/Painting/CommandExecutorCPU.h

```diff
@@ -67,10 +72,34 @@ class CommandExecutorCPU : public CommandExecutor {
         Optional<StackingContextMask> mask = {};
     };
 
-    [[nodiscard]] Gfx::Painter const& painter() const { return *stacking_contexts.last().painter; }
-    [[nodiscard]] Gfx::Painter& painter() { return *stacking_contexts.last().painter; }
+    struct ForegroundTextAlphaMask {
+        RefPtr<Gfx::Bitmap> background_canvas;
+        RefPtr<Gfx::Bitmap> mask_canvas;
+        OwnPtr<Gfx::Painter> background_painter;
+        OwnPtr<Gfx::Painter> mask_painter;
+        Gfx::IntRect destination_rect;
+    };
+
+    [[nodiscard]] Gfx::Painter const& painter() const {
+        if (m_alpha_mask_stack.size() > 0) {
+            // If rendering to an alpha mask, by default render to the background canvas.
+            return *(m_foreground_text_alpha_masks[m_alpha_mask_stack.last()]->background_painter);
+        }
+        return *stacking_contexts.last().painter;
+    }
+    [[nodiscard]] Gfx::Painter& painter() {
+        if (m_alpha_mask_stack.size() > 0) {
+            // If rendering to an alpha mask, by default render to the background canvas.
+            return *(m_foreground_text_alpha_masks[m_alpha_mask_stack.last()]->background_painter);
+        }
+        return *stacking_contexts.last().painter;
+    }
+    [[nodiscard]] Gfx::Painter const& target_painter() const { return *stacking_contexts.last().painter; }
+    [[nodiscard]] Gfx::Painter& target_painter() { return *stacking_contexts.last().painter; }
 
     Vector<StackingContext> stacking_contexts;
+    Vector<OwnPtr<ForegroundTextAlphaMask>> m_foreground_text_alpha_masks;
+    Vector<u32> m_alpha_mask_stack;
```

Comment:
Unfortunately not, at least with the approach I've gone with, the way it works is when an alpha mask is set active all non-glyph commands paint to a separate background bitmap, once all children have been painted we apply the alpha mask to the background bitmap and then paint it to the final target canvas. If all background commands  were not aware of the masking then they would paint on the target bitmap, and I don't see how we can apply the alpha mask directly to the target bitmap as the masked pixels need to be replaced with the colour of the content behind the background, this information is only known before the background is painted.

Let me know if you can think of a better way to do this.

Reply:
> Unfortunately not, at least with the approach I've gone with, the way it works is when an alpha mask is set active all non-glyph commands paint to a separate background bitmap, once all children have been painted we apply the alpha mask to the background bitmap and then paint it to the final target canvas.

what if instead of waiting for all children to be painted, a mask were applied to each DrawGlyphRun command separately?

> If all background commands were not aware of the masking then they would paint on the target bitmap, and I don't see how we can apply the alpha mask directly to the target bitmap as the masked pixels need to be replaced with the colour of the content behind the background, this information is only known before the background is painted.

what do you mean by background commands?

Reply:
> what if instead of waiting for all children to be painted, a mask were applied to each DrawGlyphRun command separately?

The mask cannot be applied when drawing the glyphs because that would mask out the background layer for the subsequent glyphs. We need to wait for all text to be painted to the alpha mask so we can then clip out the background only where there is no text.

> what do you mean by background commands?

I mean the paint commands that are not text related that contribute to the background layer that we are clipping. Commands like `fill_rect` / `draw_scaled_bitmap`.

Reply:
> I mean the paint commands that are not text related that contribute to the background layer that we are clipping.

what happens if glyph painting commands are interleaved with "background painting commands"? I build this branch to check but it crashes on this page, for example:
```html
<!DOCTYPE html>
<style>
.gradient-text {
    font-size: 60px;
    font-weight: bold;
    color: transparent;
    background-image: linear-gradient(45deg, #ff69b4, #8a2be2);
    background-clip: text;
    display: inline-block;
}
</style>
<div class="gradient-text">Gradient Text<div style="display: inline-block; opacity: 0.5">hmmm</div></div>
```

Reply:
Ahhh good catch! I clearly need to do more testing and read up on the codebase / specs more 😅 (this is my first attempt at contributing and still learning), removing line 502 in `CommandExecutorCPU.cpp` fixes the crash, but it's definitely not matching the behaviour of Chrome, plus text highlighting is a bit broken too, so will look into it. Cheers :D

This is the code I've been using for testing so far:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>

  <style>
    p, .container {
      border: 0.8em darkviolet;
      border-style: dotted double;
      margin: 1em 0;
      padding: 1.4em;
      background: linear-gradient(60deg, red, yellow, red, yellow, red);
      font: 900 1.2em sans-serif;
      text-decoration: underline;
    }

    .border-box {
      background-clip: border-box;
    }
    .padding-box {
      background-clip: padding-box;
    }
    .content-box {
      background-clip: content-box;
    }

    .text {
      background-clip: text;
      color: rgb(0 0 0 / 20%);
    }
  </style>
</head>
<body>
  <p class="border-box">The background extends behind the border.</p>
  <p class="padding-box">
    The background extends to the inside edge of the border.
  </p>
  <p class="content-box">
    The background extends only to the edge of the content box.
  </p>
  <div class="text container">
    The background is clipped to the foreground text.
    <span>Some other text in a sub-element</span>
  </div>
  <div class="text container">
    <div>
      <div>The is nested text that should still be clipped to the background</div>
    </div>
  </div>
</body>
</html>
```